# Code-Profiling
This notebook shows 3 different ways of using [line_profiler](https://github.com/rkern/line_profiler#line-profiler) to profile the runtime of Python code line by line.

## Runtime profiling with line_profiler

### Using line_profiler within session

In [1]:
from line_profiler import LineProfiler

In [2]:
def function_1(n):
    for i in range(n):
        x = [5 * j for j in range(10000)]
        y = function_2(n)
    
def function_2(n):
    s1 = [i for i in range(n)]
    s2 = sum(s1)

In [3]:
lp = LineProfiler()
lp_wrapper = lp(function_1)
lp.add_function(function_2)   # add additional function to profile
lp_wrapper(n = 1000)
lp.print_stats()

Timer unit: 1e-06 s

Total time: 1.60698 s
File: <ipython-input-2-5a497711d596>
Function: function_1 at line 1

Line #      Hits         Time  Per Hit   % Time  Line Contents
     1                                           def function_1(n):
     2      1001        446.0      0.4      0.0      for i in range(n):
     3      1000    1489773.0   1489.8     92.7          x = [5 * j for j in range(10000)]
     4      1000     116759.0    116.8      7.3          y = function_2(n)

Total time: 0.109381 s
File: <ipython-input-2-5a497711d596>
Function: function_2 at line 6

Line #      Hits         Time  Per Hit   % Time  Line Contents
     6                                           def function_2(n):
     7      1000     103805.0    103.8     94.9      s1 = [i for i in range(n)]
     8      1000       5576.0      5.6      5.1      s2 = sum(s1)



### %lprun magic function
First we must load the line_profiler Ipython extension;

In [4]:
%load_ext line_profiler

Note that the output of %lprun is displayed in a sub window in the browser so has been copied into a markdown cell here.

In [5]:
%lprun -f function_1 -f function_2 function_1(1000)

Timer unit: 1e-06 s

Total time: 1.56766 s
File: <ipython-input-2-5a497711d596>
Function: function_1 at line 1

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
     1                                           def function_1(n):
     2      1001        428.0      0.4      0.0      for i in range(n):
     3      1000    1454852.0   1454.9     92.8          x = [5 * j for j in range(10000)]
     4      1000     112383.0    112.4      7.2          y = function_2(n)

Total time: 0.105065 s
File: <ipython-input-2-5a497711d596>
Function: function_2 at line 6

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
     6                                           def function_2(n):
     7      1000      99636.0     99.6     94.8      s1 = [i for i in range(n)]
     8      1000       5429.0      5.4      5.2      s2 = sum(s1)

### From command line
It is also possible to use line_profiler from the command line. The functions above have been saved in a file 'functions_to_profile.py';

In [6]:
with open('functions_to_profile.py', 'r') as viewFileOpen:
        data = viewFileOpen.read()
print(data)

@profile
def function_1(n):
    for i in range(n):
        x = [5 * j for j in range(10000)]
        y = function_2(n)
    
@profile
def function_2(n):
    s1 = [i for i in range(n)]
    s2 = sum(s1)


function_1(1000)



Now we use the %system magic command to run a command line command in jupyter and capture the results;

In [7]:
%system kernprof -l -v functions_to_profile.py

['Wrote profile results to functions_to_profile.py.lprof',
 'Timer unit: 1e-06 s',
 '',
 'Total time: 1.53509 s',
 'File: functions_to_profile.py',
 'Function: function_1 at line 1',
 '',
 'Line #      Hits         Time  Per Hit   % Time  Line Contents',
 '==============================================================',
 '     1                                           @profile',
 '     2                                           def function_1(n):',
 '     3      1001        431.0      0.4      0.0      for i in range(n):',
 '     4      1000    1420906.0   1420.9     92.6          x = [5 * j for j in range(10000)]',
 '     5      1000     113757.0    113.8      7.4          y = function_2(n)',
 '',
 'Total time: 0.10519 s',
 'File: functions_to_profile.py',
 'Function: function_2 at line 7',
 '',
 'Line #      Hits         Time  Per Hit   % Time  Line Contents',
 '==============================================================',
 '     7                                           @pro